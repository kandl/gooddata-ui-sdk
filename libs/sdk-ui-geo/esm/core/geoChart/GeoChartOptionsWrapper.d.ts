import React from "react";
import { IGeoChartInnerProps } from "./GeoChartInner.js";
import { IColorStrategy, IPushpinCategoryLegendItem } from "@gooddata/sdk-ui-vis-commons";
export declare class GeoChartOptionsWrapper extends React.Component<IGeoChartInnerProps> {
    private readonly emptyHeaderString;
    private readonly nullHeaderString;
    private readonly errorMap;
    constructor(props: IGeoChartInnerProps);
    render(): string | number | boolean | React.ReactFragment | JSX.Element | null | undefined;
    renderVisualization(): React.ReactNode;
    private buildGeoChartOptions;
    private getCategoryLegendItems;
    private validateData;
}
export declare function createCategoryLegendItems(colorStrategy: IColorStrategy, emptyHeaderString: string, nullHeaderString: string): IPushpinCategoryLegendItem[];
//# sourceMappingURL=GeoChartOptionsWrapper.d.ts.map