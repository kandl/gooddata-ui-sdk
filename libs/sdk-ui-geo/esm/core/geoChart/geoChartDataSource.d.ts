import { IGeoConfig, IGeoData } from "../../GeoChart.js";
import { IColorStrategy } from "@gooddata/sdk-ui-vis-commons";
export interface IGeoDataSourceProps {
    colorStrategy: IColorStrategy;
    config: IGeoConfig;
    geoData: IGeoData;
    hasClustering: boolean;
}
type IGeoDataSourceFeature = GeoJSON.Feature<GeoJSON.Point, GeoJSON.GeoJsonProperties>;
export type IGeoDataSourceFeatures = IGeoDataSourceFeature[];
export declare const createPushpinDataSource: (dataSourceProps: IGeoDataSourceProps) => mapboxgl.GeoJSONSourceRaw;
export {};
//# sourceMappingURL=geoChartDataSource.d.ts.map