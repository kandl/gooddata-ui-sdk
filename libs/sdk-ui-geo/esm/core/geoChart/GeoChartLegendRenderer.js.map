{"version":3,"file":"GeoChartLegendRenderer.js","sourceRoot":"","sources":["../../../src/core/geoChart/GeoChartLegendRenderer.tsx"],"names":[],"mappings":"AAAA,qCAAqC;AACrC,OAAO,KAAK,MAAM,OAAO,CAAC;AAC1B,OAAO,YAAqD,MAAM,eAAe,CAAC;AAClF,OAAO,EAAE,MAAM,YAAY,CAAC;AAC5B,OAAO,EAAE,aAAa,EAAE,MAAM,gBAAgB,CAAC;AAE/C,OAAO,EAAE,uBAAuB,EAAE,MAAM,oBAAoB,CAAC;AAC7D,OAAO,qBAAqB,EAAE,EAAE,qBAAqB,EAAE,MAAM,oCAAoC,CAAC;AAClG,OAAO,iBAAiB,MAAM,gCAAgC,CAAC;AAE/D,OAAO,EAAE,mBAAmB,EAAE,MAAM,4BAA4B,CAAC;AACjE,OAAO,EAA4C,MAAM,EAAE,WAAW,EAAE,MAAM,8BAA8B,CAAC;AAC7G,OAAO,EACH,yBAAyB,EACzB,0BAA0B,EAC1B,wBAAwB,EACxB,yBAAyB,EACzB,mCAAmC,EACnC,wBAAwB,GAC3B,MAAM,kCAAkC,CAAC;AAE1C,MAAM,sBAAsB,GAAG,GAAG,CAAC;AAEnC,4FAA4F;AAC5F,8GAA8G;AAC9G,+EAA+E;AAC/E,MAAM,OAAO,GAAG,aAAa,CAAC,YAAY,CAAC,CAAC;AAwB5C,SAAS,aAAa,CAAC,KAAmC,EAAE,gBAAmC;IAC3F,MAAM,EAAE,QAAQ,EAAE,aAAa,EAAE,GAAG,KAAK,CAAC;IAC1C,MAAM,EAAE,aAAa,EAAE,GAAG,gBAAgB,CAAC;IAC3C,MAAM,gBAAgB,GAAG,aAAa,IAAI,QAAQ,KAAK,QAAQ,CAAC;IAChE,OAAO,EAAE,CAAC,YAAY,EAAE,cAAc,EAAE,YAAY,EAAE;QAClD,uBAAuB,EAAE,aAAa;QACtC,wBAAwB,EAAE,CAAC,aAAa;QACxC,MAAM,EAAE,CAAC,aAAa;QACtB,iBAAiB,EAAE,aAAa;QAChC,CAAC,YAAY,QAAQ,EAAE,CAAC,EAAE,CAAC,gBAAgB;QAC3C,2BAA2B;QAC3B,mDAAmD;QACnD,iBAAiB,EAAE,gBAAgB;KACtC,CAAC,CAAC;AACP,CAAC;AAED,MAAM,CAAC,OAAO,UAAU,sBAAsB,CAAC,KAAmC;IAC9E,MAAM,EAAE,aAAa,GAAG,EAAE,EAAE,OAAO,GAAG,EAAE,EAAE,MAAM,EAAE,cAAc,GAAG,EAAE,EAAE,GAAG,KAAK,CAAC;IAChF,MAAM,QAAQ,GAAG,KAAK,CAAC,QAAQ,IAAI,KAAK,CAAC;IAEzC,MAAM,gBAAgB,GAAsB,mBAAmB,CAAC,aAAa,EAAE,OAAO,CAAC,CAAC;IACxF,MAAM,EAAE,iBAAiB,EAAE,cAAc,EAAE,aAAa,EAAE,GAAG,gBAAgB,CAAC;IAC9E,MAAM,eAAe,GAAG,iBAAiB,IAAI,cAAc,IAAI,aAAa,CAAC;IAE7E,IAAI,CAAC,eAAe,EAAE;QAClB,OAAO,IAAI,CAAC;KACf;IAED,IAAI,iBAAiB,EAAE;QACnB,OAAO,2BAA2B,CAAC,KAAK,EAAE,gBAAgB,CAAC,CAAC;KAC/D;IAED,IAAI,cAAc,IAAI,aAAa,IAAI,sBAAsB,CAAC,MAAM,EAAE,QAAQ,CAAC,EAAE;QAC7E,OAAO,CACH,oBAAC,4BAA4B,OACrB,KAAK,EACT,cAAc,EAAE,cAAc,EAC9B,gBAAgB,EAAE,gBAAgB,GACpC,CACL,CAAC;KACL;IAED,OAAO,wBAAwB,CAAC,KAAK,EAAE,gBAAgB,CAAC,CAAC;AAC7D,CAAC;AAED,SAAS,2BAA2B,CAChC,KAAmC,EACnC,gBAAmC;IAEnC,MAAM,EAAE,WAAW,EAAE,WAAW,EAAE,GAAG,KAAK,CAAC;IAC3C,MAAM,EAAE,aAAa,EAAE,GAAG,gBAAgB,CAAC;IAC3C,MAAM,UAAU,GAAG,EAAE,CAAC,aAAa,CAAC,KAAK,EAAE,gBAAgB,CAAC,CAAC,CAAC;IAC9D,OAAO,CACH,oBAAC,OAAO,IAAC,MAAM,EAAE,IAAI,IAChB,CAAC,EAAE,UAAU,EAAE,WAAW,EAAE,iBAAiB,EAA0B,EAAE,EAAE;QACxE,IAAI,mCAAmC,CAAC,WAAW,EAAE,MAAM,EAAE,KAAK,EAAE,WAAW,CAAC,EAAE;YAC9E,OAAO,CACH,6BAAK,SAAS,EAAE,UAAU,EAAE,GAAG,EAAE,UAAU,IACtC,mBAAmB,CAAC,KAAK,EAAE,iBAAiB,EAAE,aAAa,CAAC,CAC3D,CACT,CAAC;SACL;QAED,OAAO,CACH,6BAAK,SAAS,EAAE,UAAU,EAAE,GAAG,EAAE,UAAU;YACtC,2BAA2B,CAAC,KAAK,EAAE,iBAAiB,EAAE,aAAa,CAAC;YACpE,uBAAuB,CAAC,KAAK,EAAE,aAAa,CAAC,CAC5C,CACT,CAAC;IACN,CAAC,CACK,CACb,CAAC;AACN,CAAC;AAED,sFAAsF;AACtF,SAAS,sBAAsB,CAAC,MAA0B,EAAE,cAA4B;IACpF,IAAI,MAAM,KAAK,SAAS,IAAI,CAAC,cAAc,KAAK,MAAM,IAAI,cAAc,KAAK,OAAO,CAAC,EAAE;QACnF,MAAM,0BAA0B,GAAG,sBAAsB,GAAG,qBAAqB,CAAC;QAClF,OAAO,MAAM,GAAG,0BAA0B,CAAC;KAC9C;IACD,OAAO,KAAK,CAAC;AACjB,CAAC;AAOD,SAAS,4BAA4B,CAAC,KAAyC;IAC3E,MAAM,CAAC,IAAI,EAAE,OAAO,CAAC,GAAG,KAAK,CAAC,QAAQ,CAAS,CAAC,CAAC,CAAC;IAElD,MAAM,YAAY,GAAG,GAAS,EAAE,CAAC,OAAO,CAAC,CAAC,CAAC,CAAC;IAC5C,MAAM,YAAY,GAAG,GAAS,EAAE,CAAC,OAAO,CAAC,CAAC,CAAC,CAAC;IAE5C,MAAM,EAAE,gBAAgB,EAAE,GAAG,KAAK,CAAC;IACnC,MAAM,UAAU,GAAG,aAAa,CAAC,KAAK,EAAE,gBAAgB,CAAC,CAAC;IAC1D,OAAO,CACH,6BAAK,SAAS,EAAE,UAAU;QACtB,6BAAK,SAAS,EAAC,mBAAmB;YAC7B,wBAAwB,CAAC,KAAK,EAAE,IAAI,KAAK,CAAC,CAAC;YAC3C,uBAAuB,CAAC,KAAK,EAAE,IAAI,KAAK,CAAC,CAAC,CACzC;QACN,oBAAC,MAAM,IAAC,IAAI,EAAE,IAAI,EAAE,UAAU,EAAE,CAAC,EAAE,YAAY,EAAE,YAAY,EAAE,YAAY,EAAE,YAAY,GAAI,CAC3F,CACT,CAAC;AACN,CAAC;AAED,SAAS,wBAAwB,CAC7B,KAAmC,EACnC,gBAAmC;IAEnC,MAAM,EAAE,cAAc,EAAE,aAAa,EAAE,GAAG,gBAAgB,CAAC;IAC3D,MAAM,UAAU,GAAG,aAAa,CAAC,KAAK,EAAE,gBAAgB,CAAC,CAAC;IAE1D,OAAO,CACH,6BAAK,SAAS,EAAE,UAAU;QACrB,wBAAwB,CAAC,KAAK,EAAE,cAAc,CAAC;QAC/C,uBAAuB,CAAC,KAAK,EAAE,aAAa,CAAC,CAC5C,CACT,CAAC;AACN,CAAC;AAED,SAAS,wBAAwB,CAC7B,KAAmC,EACnC,cAAuB;IAEvB,IAAI,CAAC,cAAc,EAAE;QACjB,OAAO,IAAI,CAAC;KACf;IAED,MAAM,EACF,OAAO,EACP,QAAQ,GAAG,KAAK,EAChB,UAAU,EACV,aAAa,EACb,gBAAgB,EAChB,cAAc,GAAG,EAAE,EACnB,WAAW,GACd,GAAG,KAAK,CAAC;IAEV,MAAM,IAAI,GAAG,OAAQ,CAAC,KAAM,CAAC,IAAI,CAAC;IAClC,MAAM,MAAM,GAAG,OAAQ,CAAC,KAAM,CAAC,MAAM,CAAC;IACtC,MAAM,eAAe,GAAG,IAAI,CAAC,MAAM,CAAC,QAAQ,CAAC,CAAC;IAC9C,MAAM,SAAS,GAAG,uBAAuB,CAAC,eAAe,EAAE,gBAAgB,CAAC,CAAC;IAC7E,MAAM,KAAK,GAAG,WAAW,EAAE,MAAM,EAAE,KAAK,CAAC;IACzC,MAAM,IAAI,GAAG,yBAAyB,CAAC,KAAK,EAAE,aAAa,EAAE,UAAU,CAAC,CAAC;IACzE,MAAM,KAAK,GAAG,0BAA0B,CAAC,OAAO,EAAE,KAAK,EAAE,IAAI,EAAE,KAAK,EAAE,UAAU,CAAC,CAAC;IAElF,OAAO,CACH,oBAAC,WAAW,IACR,IAAI,EAAE,SAAS,EACf,MAAM,EAAE,MAAM,EACd,IAAI,EAAE,IAAI,EACV,cAAc,EAAE,cAAc,EAC9B,QAAQ,EAAE,QAAQ,EAClB,KAAK,EAAE,KAAK,GACd,CACL,CAAC;AACN,CAAC;AAED,SAAS,2BAA2B,CAChC,KAAmC,EACnC,WAAwB,EACxB,aAAsB;IAEtB,MAAM,EAAE,WAAW,EAAE,GAAG,KAAK,CAAC;IAE9B,IAAI,CAAC,WAAW,EAAE;QACd,OAAO,IAAI,CAAC;KACf;IAED,OAAO,CACH,oBAAC,qBAAqB,OACd,KAAK,EACT,WAAW,EAAE,WAAW,EACxB,aAAa,EAAE,aAAa,EAC5B,WAAW,EAAE,WAAW,GAC1B,CACL,CAAC;AACN,CAAC;AAED,SAAS,uBAAuB,CAC5B,KAAmC,EACnC,aAAsB,EACtB,iBAAiB,GAAG,KAAK,EACzB,eAAe,GAAG,KAAK;IAEvB,IAAI,CAAC,aAAa,EAAE;QAChB,OAAO,IAAI,CAAC;KACf;IAED,MAAM,EAAE,OAAO,EAAE,cAAc,GAAG,EAAE,EAAE,UAAU,EAAE,WAAW,EAAE,GAAG,KAAK,CAAC;IAExE,IAAI,CAAC,OAAO,EAAE,IAAI,EAAE;QAChB,OAAO,IAAI,CAAC;KACf;IAED,MAAM,EACF,IAAI,EAAE,EAAE,IAAI,EAAE,MAAM,EAAE,IAAI,EAAE,GAC/B,GAAG,OAAO,CAAC;IACZ,MAAM,KAAK,GAAG,WAAW,EAAE,MAAM,EAAE,KAAK,CAAC;IAEzC,MAAM,OAAO,GAAG,wBAAwB,CAAC,KAAK,EAAE,eAAe,EAAE,UAAU,CAAC,CAAC;IAC7E,MAAM,KAAK,GAAG,yBAAyB,CAAC,IAAI,EAAE,KAAK,EAAE,iBAAiB,CAAC,CAAC;IACxE,MAAM,gBAAgB,GAAG,wBAAwB,CAAC,KAAK,EAAE,eAAe,CAAC,CAAC;IAE1E,OAAO,CACH,oBAAC,iBAAiB,IACd,MAAM,EAAE,MAAM,EACd,WAAW,EAAE,KAAK,EAClB,cAAc,EAAE,cAAc,EAC9B,KAAK,EAAE,IAAI,EACX,OAAO,EAAE,OAAO,EAChB,gBAAgB,EAAE,gBAAgB,GACpC,CACL,CAAC;AACN,CAAC;AAED,SAAS,mBAAmB,CACxB,KAAmC,EACnC,WAAwB,EACxB,aAAsB;IAEtB,MAAM,EAAE,WAAW,EAAE,GAAG,KAAK,CAAC;IAE9B,IAAI,CAAC,WAAW,EAAE;QACd,OAAO,IAAI,CAAC;KACf;IAED,OAAO,CACH,oBAAC,qBAAqB,OACd,KAAK,EACT,WAAW,EAAE,WAAW,EACxB,aAAa,EAAE,aAAa,EAC5B,mBAAmB,EAAE,KAAK,EAC1B,WAAW,EAAE,WAAW,EACxB,eAAe,EAAE,uBAAuB,CAAC,KAAK,EAAE,aAAa,EAAE,IAAI,EAAE,IAAI,CAAC,EAC1E,mBAAmB,EAAE,KAAK,CAAC,OAAO,EAAE,IAAI,EAAE,IAAI,GAChD,CACL,CAAC;AACN,CAAC"}