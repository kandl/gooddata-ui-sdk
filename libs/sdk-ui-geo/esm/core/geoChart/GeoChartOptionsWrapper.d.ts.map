{"version":3,"file":"GeoChartOptionsWrapper.d.ts","sourceRoot":"","sources":["../../../src/core/geoChart/GeoChartOptionsWrapper.tsx"],"names":[],"mappings":"AACA,OAAO,KAAK,MAAM,OAAO,CAAC;AAK1B,OAAO,EAAwC,mBAAmB,EAAE,MAAM,oBAAoB,CAAC;AAW/F,OAAO,EAEH,cAAc,EACd,0BAA0B,EAC7B,MAAM,8BAA8B,CAAC;AAGtC,qBAAa,sBAAuB,SAAQ,KAAK,CAAC,SAAS,CAAC,mBAAmB,CAAC;IAC5E,OAAO,CAAC,QAAQ,CAAC,iBAAiB,CAAS;IAC3C,OAAO,CAAC,QAAQ,CAAC,gBAAgB,CAAS;IAC1C,OAAO,CAAC,QAAQ,CAAC,QAAQ,CAAoB;gBAEjC,KAAK,EAAE,mBAAmB;IAO/B,MAAM;IA4BN,mBAAmB,IAAI,KAAK,CAAC,SAAS;IA0B7C,OAAO,CAAC,oBAAoB,CAmB1B;IAEF,OAAO,CAAC,sBAAsB;IAI9B,OAAO,CAAC,YAAY,CAWlB;CACL;AAED,wBAAgB,yBAAyB,CACrC,aAAa,EAAE,cAAc,EAC7B,iBAAiB,EAAE,MAAM,EACzB,gBAAgB,EAAE,MAAM,GACzB,0BAA0B,EAAE,CAe9B"}