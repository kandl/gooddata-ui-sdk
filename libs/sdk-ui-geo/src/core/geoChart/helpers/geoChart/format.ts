// (C) 2019-2020 GoodData Corporation
import { colors2Object, INumberObject, numberFormat } from "@gooddata/numberjs";
import { ISeparators } from "@gooddata/sdk-ui";
import escape from "lodash/escape.js";
import unescape from "lodash/unescape.js";

const customEscape = (str: string) => str && escape(unescape(str));

export function formatValueForTooltip(
    val: string | number,
    format: string | undefined,
    separators?: ISeparators,
): string {
    if (!format) {
        return `${val}`;
    }

    const formattedObject: INumberObject = colors2Object(numberFormat(val, format, undefined, separators));
    return customEscape(formattedObject.label);
}

export function getTooltipContentWidth(
    isFullScreenTooltip: boolean,
    chartWidth: number,
    tooltipMaxWidth: number,
): number {
    return isFullScreenTooltip ? chartWidth : Math.min(chartWidth, tooltipMaxWidth);
}
