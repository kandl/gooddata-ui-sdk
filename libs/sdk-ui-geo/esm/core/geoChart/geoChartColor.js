// (C) 2020-2022 GoodData Corporation
import range from "lodash/range.js";
import isEmpty from "lodash/isEmpty.js";
import isFinite from "lodash/isFinite.js";
import { DEFAULT_PUSHPIN_BORDER_COLOR_VALUE, DEFAULT_PUSHPIN_COLOR_OPACITY, DEFAULT_PUSHPIN_COLOR_SCALE, } from "./constants/geoChart.js";
import { getColorPalette, rgbToRgba } from "./helpers/geoChart/colors.js";
import { isAttributeDescriptor, isResultAttributeHeader } from "@gooddata/sdk-model";
import { getMinMax } from "./helpers/geoChart/common.js";
const DEFAULT_SEGMENT_ITEM = "default_segment_item";
const DEFAULT_COLOR_INDEX_IN_PALETTE = DEFAULT_PUSHPIN_COLOR_SCALE - 1;
export function getColorIndexInPalette(value, min, max) {
    if (value === null || !isFinite(value) || min === max || value === min) {
        return 0;
    }
    if (value === max) {
        return DEFAULT_COLOR_INDEX_IN_PALETTE;
    }
    const step = (max - min) / DEFAULT_PUSHPIN_COLOR_SCALE;
    for (let i = 0, offset = min; i < DEFAULT_PUSHPIN_COLOR_SCALE; i++, offset += step) {
        if (offset >= value) {
            return i;
        }
    }
    return DEFAULT_COLOR_INDEX_IN_PALETTE;
}
export function getColorPaletteMapping(colorStrategy) {
    const colorAssignment = colorStrategy.getColorAssignment();
    return colorAssignment.reduce((result, item, index) => {
        const color = colorStrategy.getColorByIndex(index);
        const colorPalette = getColorPalette(color, DEFAULT_PUSHPIN_COLOR_OPACITY);
        // color base on Location
        if (isAttributeDescriptor(item.headerItem)) {
            return {
                [DEFAULT_SEGMENT_ITEM]: colorPalette,
            };
        }
        // color base on SegmentBy
        const name = isResultAttributeHeader(item.headerItem)
            ? item.headerItem.attributeHeaderItem.name
            : DEFAULT_SEGMENT_ITEM;
        result[name ?? DEFAULT_SEGMENT_ITEM] = colorPalette;
        return result;
    }, {});
}
/**
 * Return RGB border and background colors base on color and segment values
 *
 * Example:
 * ```
 *  [any-number] => [{
 *       border: "rgb(127,224,198)",
 *       background: "rgb(215,242,250)",
 *  }]
 * ```
 */
export function getPushpinColors(colorValues, segmentValues = [], colorStrategy) {
    const defaultColorValue = colorStrategy.getColorByIndex(0);
    const defaultColor = rgbToRgba(defaultColorValue, DEFAULT_PUSHPIN_COLOR_OPACITY);
    if (!colorValues.length && !segmentValues.length) {
        return [
            {
                border: DEFAULT_PUSHPIN_BORDER_COLOR_VALUE,
                background: defaultColor,
            },
        ];
    }
    const segmentNames = [...segmentValues];
    const colorPaletteMapping = getColorPaletteMapping(colorStrategy);
    if (!colorValues.length) {
        return segmentNames.map((name) => {
            const palette = colorPaletteMapping[name];
            return {
                border: DEFAULT_PUSHPIN_BORDER_COLOR_VALUE,
                background: palette[DEFAULT_COLOR_INDEX_IN_PALETTE],
            };
        });
    }
    const colorsWithoutNull = colorValues.filter((value) => value !== null && isFinite(value));
    const { min, max } = getMinMax(colorsWithoutNull);
    if (min === max && !segmentValues.length) {
        return [
            {
                border: DEFAULT_PUSHPIN_BORDER_COLOR_VALUE,
                background: defaultColor,
            },
        ];
    }
    return colorValues.map((color, index) => {
        const value = color !== null && isFinite(color) ? color : min;
        const colorIndex = getColorIndexInPalette(value, min, max);
        const segmentItemName = segmentNames[index] || DEFAULT_SEGMENT_ITEM;
        const palette = colorPaletteMapping[segmentItemName];
        return {
            border: palette[DEFAULT_COLOR_INDEX_IN_PALETTE],
            background: palette[colorIndex],
        };
    });
}
export function generateLegendColorData(colorSeries, colorString) {
    if (isEmpty(colorSeries)) {
        return [];
    }
    const colorPalette = getColorPalette(colorString, DEFAULT_PUSHPIN_COLOR_OPACITY);
    const min = Math.min(...colorSeries);
    const max = Math.max(...colorSeries);
    const offset = (max - min) / DEFAULT_PUSHPIN_COLOR_SCALE;
    if (min === max) {
        return [];
    }
    return range(0, DEFAULT_PUSHPIN_COLOR_SCALE).map((index) => {
        const from = min + offset * index;
        const isLastItem = index === DEFAULT_PUSHPIN_COLOR_SCALE - 1;
        const to = isLastItem ? max : from + offset;
        const range = {
            from,
            to,
        };
        return {
            range,
            color: colorPalette[index],
        };
    });
}
//# sourceMappingURL=geoChartColor.js.map