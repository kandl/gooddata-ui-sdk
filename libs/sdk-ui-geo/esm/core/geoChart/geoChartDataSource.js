// (C) 2019-2022 GoodData Corporation
import { getPushpinColors } from "./geoChartColor.js";
import { DEFAULT_CLUSTER_RADIUS, DEFAULT_CLUSTER_MAX_ZOOM, PUSHPIN_SIZE_OPTIONS_MAP, } from "./constants/geoChart.js";
import { getMinMax } from "./helpers/geoChart/common.js";
function transformPushpinDataSource(dataSourceProps) {
    const { config, geoData, colorStrategy } = dataSourceProps;
    const { color, location, segment, size, tooltipText } = geoData;
    if (!location) {
        /*
         * short-circuit the processing at this point. if there is no location data then the features will
         * be empty
         */
        return [];
    }
    const { points: geoPointsConfig = {} } = config || {};
    const locationNameTitle = tooltipText ? tooltipText.name : "";
    const colorTitle = color ? color.name : "";
    const sizeTitle = size ? size.name : "";
    const segmentTitle = segment ? segment.name : "";
    const hasSize = size !== undefined;
    const locationData = location.data;
    const locationNameData = tooltipText?.data ?? [];
    const segmentData = segment?.data ?? [];
    const segmentUris = segment?.uris ?? [];
    const sizeData = size?.data ?? [];
    const { min: minSizeFromData, max: maxSizeFromData } = getMinMax(sizeData);
    const colorData = color?.data ?? [];
    const sizeFormat = size?.format ?? "";
    const colorFormat = color?.format ?? "";
    const pushpinColors = getPushpinColors(colorData, segmentData, colorStrategy);
    return locationData.reduce((result, coordinates, index) => {
        if (!coordinates) {
            return result;
        }
        const { lat, lng } = coordinates;
        const pushpinSize = hasSize
            ? calculateSizeInPixel(sizeData[index], minSizeFromData, maxSizeFromData, geoPointsConfig)
            : PUSHPIN_SIZE_OPTIONS_MAP.min.default;
        const colorValue = colorData[index];
        const segmentValue = segmentData[index];
        const segmentUri = segmentUris[index];
        const pushpinColor = pushpinColors[index] || pushpinColors[0] || {};
        result.push({
            type: "Feature",
            geometry: {
                type: "Point",
                coordinates: [lng, lat], // Mapbox requires number[]
            },
            properties: {
                pushpinSize,
                locationName: {
                    title: locationNameTitle,
                    value: locationNameData[index],
                },
                locationIndex: index,
                color: {
                    ...pushpinColor,
                    title: colorTitle,
                    value: colorValue,
                    format: colorFormat,
                },
                size: {
                    title: sizeTitle,
                    value: sizeData[index],
                    format: sizeFormat,
                },
                segment: {
                    title: segmentTitle,
                    value: segmentValue,
                    uri: segmentUri,
                },
            },
        });
        return result;
    }, []);
}
// transform data value to pushpin size in pixel
const calculateSizeInPixel = (dataValue, minSize, maxSize, geoPointsConfig) => {
    if (minSize === maxSize || dataValue === null) {
        return PUSHPIN_SIZE_OPTIONS_MAP.min.default;
    }
    const { minSize: minSizeFromConfig = "default", maxSize: maxSizeFromConfig = "default" } = geoPointsConfig || {};
    const minSizeInPixel = PUSHPIN_SIZE_OPTIONS_MAP.min[minSizeFromConfig];
    const maxSizeInPixel = PUSHPIN_SIZE_OPTIONS_MAP.max[maxSizeFromConfig];
    return ((dataValue - minSize) * (maxSizeInPixel - minSizeInPixel)) / (maxSize - minSize) + minSizeInPixel;
};
export const createPushpinDataSource = (dataSourceProps) => {
    const { hasClustering } = dataSourceProps;
    const source = {
        type: "geojson",
        data: {
            type: "FeatureCollection",
            features: transformPushpinDataSource(dataSourceProps),
        },
    };
    if (hasClustering) {
        return {
            ...source,
            cluster: true,
            clusterMaxZoom: DEFAULT_CLUSTER_MAX_ZOOM,
            clusterRadius: DEFAULT_CLUSTER_RADIUS,
        };
    }
    return source;
};
//# sourceMappingURL=geoChartDataSource.js.map