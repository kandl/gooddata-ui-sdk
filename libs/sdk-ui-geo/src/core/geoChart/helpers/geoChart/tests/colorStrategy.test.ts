// (C) 2020 GoodData Corporation

import { getColorStrategy } from "../../../colorStrategy/geoChart.js";
import { DefaultColorPalette } from "@gooddata/sdk-ui";
import { RecShortcuts } from "../../../../../../__mocks__/recordings.js";

describe("getColorStrategy", () => {
    it("should return GeoChartColorStrategy", () => {
        const { dv, geoData } = RecShortcuts.LocationOnly;
        const geoChartColorStrategy = getColorStrategy(DefaultColorPalette, undefined as any, geoData, dv);

        expect(geoChartColorStrategy.getColorAssignment()).toMatchSnapshot();
        expect(geoChartColorStrategy.getColorByIndex(0)).toEqual("rgb(20,178,226)");
        expect(geoChartColorStrategy.getColorByIndex(1)).toBeUndefined();
        expect(geoChartColorStrategy.getFullColorAssignment()).toMatchSnapshot();
    });
});
