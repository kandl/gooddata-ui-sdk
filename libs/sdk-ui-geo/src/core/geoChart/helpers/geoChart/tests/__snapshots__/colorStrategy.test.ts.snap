// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getColorStrategy > should return GeoChartColorStrategy 1`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "1",
    },
    "headerItem": {
      "attributeHeader": {
        "formOf": {
          "identifier": "attr.uscities.city",
          "name": "City",
          "ref": {
            "uri": "/gdc/md/liveexamplesworkspace/obj/9458",
          },
          "uri": "/gdc/md/liveexamplesworkspace/obj/9458",
        },
        "identifier": "label.uscities.city.location",
        "localIdentifier": "a_label.uscities.city.location",
        "name": "location",
        "ref": {
          "uri": "/gdc/md/liveexamplesworkspace/obj/9459",
        },
        "type": "GDC.geo.pin",
        "uri": "/gdc/md/liveexamplesworkspace/obj/9459",
      },
    },
  },
]
`;

exports[`getColorStrategy > should return GeoChartColorStrategy 2`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "1",
    },
    "headerItem": {
      "attributeHeader": {
        "formOf": {
          "identifier": "attr.uscities.city",
          "name": "City",
          "ref": {
            "uri": "/gdc/md/liveexamplesworkspace/obj/9458",
          },
          "uri": "/gdc/md/liveexamplesworkspace/obj/9458",
        },
        "identifier": "label.uscities.city.location",
        "localIdentifier": "a_label.uscities.city.location",
        "name": "location",
        "ref": {
          "uri": "/gdc/md/liveexamplesworkspace/obj/9459",
        },
        "type": "GDC.geo.pin",
        "uri": "/gdc/md/liveexamplesworkspace/obj/9459",
      },
    },
  },
]
`;
